### PR TITLE
Remove APEX

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
This patch removes (most) of our dependency on APEX by re-implementing the important features.

This was done by:
- Killing the FusedLayerNorm usage. Historically, there was a ~10% speed difference IIRC, but Pytorch has improved and closed the gap to ~2% these days last I heard. This is sufficient for cleaner, simpler code.
- Copying and re-adapting Fairseq's apex-style fp16 optimizer. It has high compatibility with the mem efficient one.

We still have the dependency for FusedAdam. This will be removed in subsequent PRs where we introduce fairscale.

Additionally, there's a small bugfix setting tolerance to 0.0 of the DynamicLossScalar. This was affecting @hadasah and occasionally causing overflows to be ignored in long jobs. Upon investigation, it turns out the default in the _class_ in fairseq is 0.05, but the default in the CLI is 0.0. We move to the better default.

**Testing steps**
Manual testing. CI.

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -1004,31 +1002,30 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
...
                 # this is a bit clunky, but alternatives are worse
                 try:
-                    self.optimizer.optimizer.load_state_dict(optim_states)
+                    self.optimizer.load_state_dict(optim_states)
                 except ValueError:
                     warn_once(
                         'WARNING: not loading optim state since model params changed.'
                     )
-                return
+                return True
```
Comment (EricMichaelSmith) [inline parlai/core/torch_agent.py:1028]: nit: for clarification, maybe it'd be useful to add in the docstring what a return value of True or False indicates?

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -1907,6 +1909,11 @@ def upgrade_opt(cls, opt_from_disk: Opt):
         # call the parent upgrades
         opt_from_disk = super(TorchAgent, cls).upgrade_opt(opt_from_disk)
 
+        if 'fp16_impl' in opt_from_disk:
+            # 2021-02-23: we removed apex and replaced it with our own thing
+            if opt_from_disk['fp16_impl'] == 'apex':
```
Comment (EricMichaelSmith) [inline parlai/core/torch_agent.py:1914]: 👍

File: parlai/core/torch_generator_agent.py [hunk 3]
```diff
@@ -628,6 +630,7 @@ def _init_cuda_buffer(self, batchsize, maxlen, force=False):
 
         This is also used in distributed mode to force a worker to sync with others.
         """
+        return
```
Comment (EricMichaelSmith) [inline parlai/core/torch_generator_agent.py:633]: Hmm I assume this is a line added for debugging?

**Review by EricMichaelSmith — APPROVED** [review]
Seems reasonable - a few comments

File: parlai/agents/transformer/modules.py [hunk 4]
```diff
@@ -29,33 +29,9 @@
 from parlai.utils.misc import warn_once
 from parlai.utils.torch import neginf, PipelineHelper
 
-try:
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/agents/transformer/modules.py:32]: im so happy right now

File: parlai/core/torch_generator_agent.py (hunk 3 above)
Comment (klshuster) [reply parlai/core/torch_generator_agent.py:633]: speaking of, can we possibly make this a flag to turn off, i.e. once we make this `debug` mode? i really would appreciate being able to turn this off when debugging...

File: parlai/utils/fp16.py [hunk 5]
```diff
@@ -85,68 +87,186 @@ def has_overflow(grad_norm):
     return False
 
 
-###################################################
-## APEX Wrapper
-###################################################
+class SafeFP16Optimizer(torch.optim.Optimizer):
```
Comment (klshuster) [inline parlai/utils/fp16.py:90]: what exactly is a safe fp16 optimizer?

File: parlai/utils/fp16.py [hunk 6]
```diff
@@ -85,68 +87,186 @@ def has_overflow(grad_norm):
...
+            overflow = has_overflow(grad_norm)
+            prev_scale = self.scaler.loss_scale
+            self.scaler.update_scale(overflow)
+            if overflow:
+                self.zero_grad()
+                if self.scaler.loss_scale <= self.min_loss_scale:
+                    # Use FloatingPointError as an uncommon error that parent
+                    # functions can safely catch to stop training.
+                    self.scaler.loss_scale = prev_scale
+                    raise FloatingPointError(
```
Comment (klshuster) [inline parlai/utils/fp16.py:222]: thanks for adding this

File: parlai/utils/fp16.py [hunk 7]
```diff
@@ -245,7 +365,7 @@ class MemoryEfficientFP16Optimizer(torch.optim.Optimizer):
     <https://github.com/pytorch/fairseq/blob/master/fairseq/optim/fp16_optimizer.py#L382>
 
     This allows you to train bigger models on a single GPU, but can be unstable.
-    Opt for the APEX implementation if you do not have concerns about memory.
+    Opt for the Pytorch implementation if you do not have concerns about memory.
```
Comment (klshuster) [inline parlai/utils/fp16.py:368]: can we say explicitly here `Safe` as well?

File: tests/test_apex.py [hunk 8]
```diff
@@ -4,21 +4,12 @@
...
-
-
 import unittest
 from parlai.core.agents import create_agent
 from parlai.core.params import ParlaiParser
 import parlai.utils.testing as testing_utils
 
 
-@unittest.skipIf(APEX_AVAILABLE, "Apex is installed, can't test its absence.")
 class TestNoApex(unittest.TestCase):
```
Comment (klshuster) [inline tests/test_apex.py:13]: why not just delete this test?

**Review by klshuster — APPROVED** [review]
🥳

File: parlai/core/torch_generator_agent.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/torch_generator_agent.py:633]: On the roadmap to get rid of it entirely

File: parlai/utils/fp16.py (hunk 5 above)
Comment (stephenroller) [reply parlai/utils/fp16.py:90]: same as the apex fp16 optimizer, just ours

File: tests/test_apex.py (hunk 8 above)
Comment (stephenroller) [reply tests/test_apex.py:13]: bc i abuse it and use it for the safe now. should rename it.

Comment (stephenroller) [reply tests/test_apex.py:13]: er, actually no. I keep it for backwards compatibility checks

but this test should be running for everyone after removing apex

**Pull request merged**